tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionParser.cs

```diff
@@ -470,7 +470,13 @@ private void AddLocalAndArgs(ScopeMember[] argsOrLocals, List<ParameterExpressio
             }
 
             var argOrLocal = argsOrLocals[index];
-            var variable = Expression.Variable(argOrLocal.Type, argOrLocal.Name);
+
+            // CRITICAL: Use runtime type (Value.GetType()) instead of declared type (Type).
+            // The compilation must use the same types as runtime.
+            // If declared type is Int32 but actual value is Int64, using declared type
+            // would cause InvalidCastException when the lambda is executed.
+            var runtimeType = argOrLocal.Value?.GetType() ?? argOrLocal.Type;
+            var variable = Expression.Variable(runtimeType, argOrLocal.Name);
```

Comment:
We compile the expression only once and reuse it, so the type we end up getting in this line might differ.
This is a general issue we have with our expressions - we compile them using the args/locals that happened to be in there in the first invocation but that might (and usually is) different from subsequent invocations.
I'm not suggesting to do a revamp right now, but as a note to keep in mind.